pkg/informant/dispatcher.go

```diff
@@ -0,0 +1,279 @@
+
+	// We do not need to close the response body according to docs.
+	// Doing so causes memory bugs.
+	c, _, err := websocket.Dial(ctx, addr, nil) //nolint:bodyclose // see comment above
+	if err != nil {
+		return disp, fmt.Errorf("error creating dispatcher: %w", err)
+	}
+
+	// Figure out protocol version
+	err = wsjson.Write(
+		ctx,
+		c,
+		api.VersionRange[api.MonitorProtoVersion]{
+			Min: api.MonitorProtoV1_0,
+			Max: api.MonitorProtoV1_0,
+		},
+	)
+	if err != nil {
+		return Dispatcher{}, fmt.Errorf("error sending protocol range to monitor: %w", err)
+	}
+	var version api.MonitorProtocolResponse
+	err = wsjson.Read(ctx, c, &version)
+	if err != nil {
+		return Dispatcher{}, fmt.Errorf("error reading monitor response during protocol handshake: %w", err)
+	}
+	if version.Error != nil {
+		return Dispatcher{}, fmt.Errorf("monitor returned error during protocol handshake: %s", *version.Error)
+	}
+	logger.Info("negotiated protocol with monitor", zap.String("protocol", version.Version.String()))
+
+	disp = Dispatcher{
+		conn:         c,
+		notifier:     notifier,
+		waiters:      make(map[uint64]util.SignalSender[*MonitorResult]),
+		counter:      0,
+		logger:       logger.Named("dispatcher"),
+		protoVersion: version.Version,
+	}
+	return disp, nil
+}
+
+// Send a message down the connection. Only call this method with types that
+// SerializeInformantMessage can handle.
+func (disp *Dispatcher) send(ctx context.Context, id uint64, message any) error {
+	data, err := api.SerializeInformantMessage(message, id)
+	if err != nil {
+		return fmt.Errorf("error serializing message: %w", err)
+	}
+	// wsjson.Write serializes whatever is passed in, and go serializes []byte
+	// by base64 encoding it, so use RawMessage to avoid serializing to []byte
+	// (done by SerializeInformantMessage), and then base64 encoding again
+	raw := json.RawMessage(data)
+	return wsjson.Write(ctx, disp.conn, &raw)
+}
+
+// Make a request to the monitor. The dispatcher will handle returning a response
+// on the provided SignalSender. The value passed into message must be a valid value
+// to send to the monitor. See the docs for SerializeInformantMessage.
+func (disp *Dispatcher) Call(ctx context.Context, sender util.SignalSender[*MonitorResult], message any) error {
+	id := disp.counter
+	disp.counter += 1
+	err := disp.send(ctx, id, message)
+	if err != nil {
+		disp.logger.Error("failed to send message", zap.Any("message", message), zap.Error(err))
+	}
+	disp.waiters[id] = sender
+	return nil
+}
+
+// Handle messages from the monitor. Make sure that all message types the monitor
+// can send are included in the inner switch statement.
+func (disp *Dispatcher) HandleMessage(
+	ctx context.Context,
+	logger *zap.Logger,
+	handleUpscaleRequest func(api.UpscaleRequest),
+	handleUpscaleConfirmation func(api.UpscaleConfirmation, uint64) error,
+	handleDownscaleResult func(api.DownscaleResult, uint64) error,
+	handleMonitorError func(api.InternalError, uint64) error,
```

Comment:
is there a reason not to make these methods on `*Dispatcher` instead of passing them in?

Reply:
My initial thought was that it forces them to be in that specific scope so that all the handlers are easy to find. I tried to design the `Dispatcher` (soon to be renamed :)?) so that it would be hard to forget or accidentally add a message type, since we don't have exhaustiveness checking for enums like in rust. That's why all the repetition is grouped together and self contained (`SerializeInformantMessage` and `HandleMessage` are the two "hot spots").